### PR TITLE
Add ignorechans to iqrm_mask

### PR DESCRIPTION
**Pull request by kmjc**
Add an optional `ignorechans` argument to `iqrm_mask`
This is useful when you have a pre-existing list of bad channels.
Votes from these channels will be ignored.

### CODE REVIEW

**Comment by kmjc** [comment]
Hi! 
I work on CHIME data and I found this really useful. It gives more consistent results than setting bad channels to some high value in whatever statistic you're using.

**Comment by v-morello** [comment]
Hi @kmjc,

I had a look and I think that the net effect of your code change is that all channels in `ignorechans` are forcibly masked, on top of those already masked by IQRM. Therefore, instead of manually editing the votes graph (which is a hack), you could simply create a fixed boolean channel mask from your `ignorechans` list, and fuse it with the mask returned by iqrm with a logical OR.

From a user perspective, I mean this:
```python

def chanlist_to_mask(chanlist, nchan):
    # TODO: check that no index in chanlist is > nchan - 1
    mask = np.zeros(nchan, dtype=bool)
    mask[chanlist] = True
    return mask

final_mask = iqrm_mask(<params>) | chanlist_to_mask(ignorechans, nchan)
```

Is that your actual intent here ? If not, please let me know what you're trying to do. If yes, then I would argue that the above is the cleaner way to achieve the result and that the present MR is unnecessary. I'd still be open to the idea of exposing an extra function `iqrm_mask_ignorechans()` (better name suggestion welcome!) that does the above in spirit, without touching the `iqrm_mask()` function itself.

In any case, thanks for contributing! :+1:

**Comment by kmjc** [comment]
Ah apologies, I did not make my motivations clear!
My aim is to ignore any votes cast by channels in ignorechans.

Our band is particularly plagued by RFI and there are a few large stretches of bad channels. I was finding that `iqrm_mask` doesn't always behave well around the edges of these areas, or in small sections of good channels which are surrounded by large stretches of bad ones.

Here's a plot showing the difference. 
iqrm_example
(In the plots stat has been masked with the ignorechans and the orange indicates channels zapped by `iqrm_mask`)
It is a purposefully an extreme case as in the top two plots I didn't fill in the known bad values with anything sensible. But if you do fill them in, what `iqrm_mask` zaps changes at the borders dependent on what your data look like and what method you use to find a fill value. I was finding the filling scheme then needed a lot of tweaking for different days and pointings.
(I think in this case I could solve the issue by increasing the radius used - it's at 10% of the band - but the stat in these plots happens to be uncommonly flat across our band. Others aren't at all, and the band has a lot of structure so I don't think that would go well)

It's definitely a hack(!), and if this seems like a niche case then no problem, please close the PR, I just thought it might be useful for other people

**Comment by v-morello** [comment]
Oh this is more interesting than I initially thought.

> My aim is to ignore any votes cast by channels in ignorechans.

So it probably means that "stat" has a low value in all of those channels ? What does "stat" represent here ? Spectral standard deviation, or something else ?
The main fundamental limitation of IQRM is that it is only as good as the contamination statistic you provide it.

Your case might be relatively common so now I'm curious. Would you mind emailing me the "stat" array you used in that example ? (vmorello, at, gmail, com). In fact, feel free to send other edge cases too. I'd like to understand for myself what's going on internally here.

**Comment by v-morello** [comment]
If you're trying to work around having wide intervals of channels having a low stat value, what you're doing is not a hack and you've got the right idea. That being said, I believe that a more efficient way of achieving the result you're after would be this below. No need to edit the votes graph, and only two lines of code changed in total, just in the "final vote count" loop:

```python
ignorechans = set(ignorechans)  # convert to set for O(1) lookup

for i, casters in votes_received.items():
    for j in casters:
        if j in votes_cast and len(votes_cast[j]) < len(votes_received[i]) and j not in ignorechans:
            mask[i] = True
            break
```

But then, it's not quite 100% satisfying yet, because the values in ignorechans are still included in the estimation of the maximum acceptable "lagged diff" values in `outlier_mask()` and it's still not as good as truly ignoring them in the whole process. We could come up with something even better.

**Comment by v-morello** [comment]
Looks like we've got ourselves a really good MR now :smile:

**Comment by kmjc** [comment]
Hi!
I ended up filling the ignorechans with NaNs and then changing the percentile computations to nanpercentile instead, which I think accomplishes this.

I've been using it for a while and it seems to work well. Would this be a good addition for iqrm at this point?

**Comment by v-morello** [comment]
Hi @kmjc, I completely forgot about this. I tried your code on a few made up test cases and it works really well. Excluding the ignored channels from the dynamic threshold calculation is the right thing to do. The potential edge cases of introducing NaNs in the data make me uncomfortable (e.g. what happens when all your lagged differences end up being NaN), but even then the code seems to behave. 

High-level summary: I am happy to merge this and it is overall a good addition. If I find some time in the future, I would like to add more test cases (i.e. get more confidence than "seems to behave" :smile:) and if necessary, make further adjustments to what you propose here.

My main comment is that the way you edit the votes dictionaries is incorrect, because it breaks the symmetry between votes cast and votes received. Doing it the correct way does not change the behaviour of "ignorechans" on the test cases I tried, so that's good news. With that being said, I also noticed on said test cases that leaving the votes dictionaries untouched and just forcing `mask[ignorechans] = True` achieves the exact same result, and significantly faster I suspect. It would take more testing to confirm this hypothesis though, so I suggest we go with modifying the votes dictionaries "the right way". You might want to test whether that still works on some reference data you have.

Code comments in my next message.

File: iqrm/core.py [hunk 1]
```diff
@@ -54,6 +54,8 @@ def iqrm_mask(x, radius=5, threshold=3.0):
         is 10% of the number of frequency channels
     threshold : float, optional
         Flagging threshold in number of Gaussian sigmas
+    ignorechans: list or ndarray
+        Channels which are known to be "bad" already. Their votes will be ignored
```
Comment (v-morello) [inline iqrm/core.py]: Their votes aren't ignored in the implementation, I would just say "Channels which are known to be bad already"

File: iqrm/core.py [hunk 2]
```diff
@@ -64,6 +66,7 @@ def iqrm_mask(x, radius=5, threshold=3.0):
         and the values are the set of array indices that received a vote from i.
     """
     x = np.asarray(x)
+    x[ignorechans] = np.nan
```
Comment (v-morello) [inline iqrm/core.py]: Mutating user input is frowned upon in general, can you change the first line to `x = np.asarray(x, copy=True)` ?

File: iqrm/core.py [hunk 3]
```diff
@@ -93,6 +96,14 @@ def iqrm_mask(x, radius=5, threshold=3.0):
             votes_cast[j].add(i)
             votes_received[i].add(j)
 
+    # Ignore some channels
+    nchans = len(x)
+    for c in ignorechans:
+        other_chans = list(range(0,c)) + list(range(c+1, nchans))
```
Comment (v-morello) [inline iqrm/core.py]: Please use `set` here because
1. The votes dicts are already dicts of sets
2. `set` has **much** faster lookups in the "vote tally" loop below, the speed difference might be dramatic for large nchans.

```
other_chans = set(range(0,c)) | set(range(c+1, nchans))
```

File: iqrm/core.py [hunk 4]
```diff
@@ -93,6 +96,14 @@ def iqrm_mask(x, radius=5, threshold=3.0):
...
+        votes_received[c] = other_chans
+        votes_cast[c] = other_chans
```
Comment (v-morello) [inline iqrm/core.py]: Both these edits to the votes dictionaries are "asymmetrical". The correct way is:

```python
        votes_received[c] = other_chans
        for i in other_chans:
            votes_cast[i].add(c)

        votes_cast[c] = other_chans
        for i in other_chans:
            votes_received[i].add(c)
```

File: iqrm/core.py [hunk 5]
```diff
@@ -93,6 +96,14 @@ def iqrm_mask(x, radius=5, threshold=3.0):
             votes_cast[j].add(i)
```
Comment (v-morello) [inline iqrm/core.py:98]: Can I ask you a favour and add the following lines just after `for i, j in zip(I, J):` 
```
i = int(i)
j = int(j)
```

I noticed that otherwise the votes dictionaries contain a mix of Python `int` and `np.int64` and I think that might cause issues. It's definitely my bad though :smile:

**Review by v-morello — CHANGES_REQUESTED** [review]

File: iqrm/core.py (hunk 4 above)
Comment (v-morello) [reply iqrm/core.py]: As I said in my first message, just not touching the votes dictionaries and setting `mask[ignorechans] = True` could be much faster (and simpler!), but you'd have to confirm whether that works. I leave that decision to you, and it can be done in a future PR.

File: iqrm/core.py (hunk 2 above)
Comment (kmjc) [reply iqrm/core.py]: I changed it to 
`x = np.array(x)`
which makes a copy by default.
I apparently had an older version of numpy in my iqrm environment and so `copy=True` threw an error. I'm happy to change it to `copy=True` if that's your preference, but this way you don't have to add in a numpy version requirement

**Comment by kmjc** [comment]
Ah yes I thought using the NaNs probably made messing with the votes redundant but I wasn't quite sure.
I did some tests and, while no means exhaustive, I did get the same resulting mask when just using the nans and the only votes relating to the bad channels changes. So I think that's the way to go!

File: iqrm/core.py (hunk 2 above)
Comment (v-morello) [reply iqrm/core.py]: That's good, thanks

**Comment by v-morello** [comment]
Hi @kmjc, sorry for the delay, I was away at a conference last week. Glad to see that the simple solution works, I'm happy to reconsider the vote adjustment later if you find out it's better on some other cases.

Happy to merge, thanks for contributing :smile:
